File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/AttEdge.java
package ru.citeck.ecos.records3.record.op.atts.service.value;

import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.commons.data.MLText;
import ru.citeck.ecos.records3.record.op.atts.dto.CreateVariant;

import java.util.List;

public interface AttEdge {

    default boolean isProtected() throws Exception {
        return false;
    }

    default boolean isMultiple() throws Exception {
        return true;
    }

    default boolean isAssociation() throws Exception {
        return false;
    }

    default boolean isSearchable() throws Exception {
        return true;
    }

    default boolean isUnreadable() throws Exception {
        return false;
    }

    @Nullable
    default List<?> getOptions() throws Exception {
        return null;
    }

    @Nullable
    default List<?> getDistinct() throws Exception {
        return null;
    }

    @Nullable
    default List<CreateVariant> getCreateVariants() throws Exception {
        return null;
    }

    @Nullable
    default Class<?> getJavaClass() throws Exception {
        return null;
    }

    @Nullable
    default String getEditorKey() throws Exception {
        return null;
    }

    @Nullable
    default String getType() throws Exception {
        return null;
    }

    @Nullable
    default MLText getTitle() throws Exception {
        return null;
    }

    @Nullable
    default MLText getDescription() throws Exception {
        return null;
    }

    @Nullable
    default String getName() throws Exception {
        return null;
    }

    @Nullable
    default Object getValue() throws Exception {
        return null;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/value/AttEdge.kt
package ru.citeck.ecos.records3.record.op.atts.service.value

import ru.citeck.ecos.commons.data.MLText
import ru.citeck.ecos.records3.record.op.atts.dto.CreateVariant

interface AttEdge {

    @Throws(Exception::class)
    fun isProtected(): Boolean = false

    @Throws(Exception::class)
    fun isMultiple(): Boolean = true

    @Throws(Exception::class)
    fun isAssociation(): Boolean = false

    @Throws(Exception::class)
    fun isSearchable(): Boolean = true

    /**
     * Can client read value of this attribute or not.
     */
    @Throws(Exception::class)
    fun canBeRead(): Boolean = true

    @Throws(Exception::class)
    fun getOptions(): List<*>? = null

    @Throws(Exception::class)
    fun getDistinct(): List<*>? = null

    @Throws(Exception::class)
    fun getCreateVariants(): List<CreateVariant>? = null

    @Throws(Exception::class)
    fun getJavaClass(): Class<*>? = null

    @Throws(Exception::class)
    fun getEditorKey(): String? = null

    /**
     * Type of attribute.
     */
    @Throws(Exception::class)
    fun getType(): String? = null

    @Throws(Exception::class)
    fun getTitle(): MLText? = null

    @Throws(Exception::class)
    fun getDescription(): MLText? = null

    @Throws(Exception::class)
    fun getName(): String? = null

    @Throws(Exception::class)
    fun getValue(): Any? = null
}
